Proposal: Add TypeScript

DukeFerdinand opened the pull request:
## :mag_right:  PR Summary
This is a very copy-pasty PR, as something like 90% of all the feature (method, class) lines are unchanged beyond some Prettier code formatting.

Here's what to look out for:
- New TS endings for all major files
- New type declarations for some of the method declarations
- New Keva namespaced types

## :heavy_check_mark: How to test
There are two things to test:
- `npm run test` - normal, single run test
- `npm run dev` - runs jest in `--watch` mode. run this for active development sessions

EDIT: From PR feedback you should also consider
- `npm run format` - manually format all of your files

No new features have been added, just the addition of TypeScript. All tests should pass as expected and nothing should error or log strange things :smile: 

## :bow: Feedback
I took some pretty big liberties here, primarily in the `Keva` type namespacing, build directory, ts-jest addition, and code formatting.

If you aren't satisfied with anything here, _please_ let me know (especially the code formatting). I'm very interested in setting up a contribution code-formatting guideline to keep us consistent as we both work on this library, but I also don't really care what that format looks like. As long as we can set it up using an auto-tool like Prettier, I'm happy :wink:

kevacoin-project commented on types/index.ts:
```diff
@@ -0,0 +1,28 @@
+  export interface Transaction {
+    tx_hash: string;
+    // FIXME: This is probably the wrong type for `n`
+    n: string;
```
The value of `n` looks like this:
```
n: [ 'Nfw2WYkGoSKve74cCfEum67x8bFgpHygxg', 1 ]
```
It is an array type, the first element is the namespace Id, while the second element is the position of the transaction (which contains the namespace) in the block.

So indeed it is not a string.

kevacoin-project commented:
Looks great to me, thanks for the PR - it is very helpful! I just have one comment about the type of `n`. I like the format change you made. Is it possible to integrate "Prettier" as an auto-tool?

DukeFerdinand commented:
@kevacoin-project I fixed the type of `n` to be that sort of tuple structure. It's now typed `n: [string, number]` :)

Also yes! I have prettier set to a `format` or `fmt` command in all of my repos. I've gone ahead and added that to the scripts in package.json, as well as adding a `.prettierrc`. If you like to add "run \<tool\> on save" features to your editor, prettier also has plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode), Sublime, and all Jetbrains tools (Webstorm, IntelliJ, etc.).

Please feel free to add any changes to that in case you don't like the choices I've made (I just threw together some of my main preferences)

kevacoin-project commented:
Excellent, thank you!
